src/app/SimpleSubscriptionResumptionStorage.cpp

```diff
@@ -0,0 +1,589 @@
+/*
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the CHIP CASE Session object that provides
+ *      APIs for constructing a secure session using a certificate from the device's
+ *      operational credentials.
+ */
+
+#include <app/SimpleSubscriptionResumptionStorage.h>
+
+#include <lib/support/Base64.h>
+#include <lib/support/SafeInt.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+namespace chip {
+namespace app {
+
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kFabricIndexTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kPeerNodeIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kSubscriptionIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kMinIntervalTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kMaxIntervalTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kFabricFilteredTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kPathCountTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kPathTypeTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kEndpointIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kClusterIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kAttributeIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kEventIdTag;
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::SaveIndex(const SubscriptionIndex & index)
```

Comment:
I think it boils down to how big the entries are. It would require quite a large heap buffer to load a big set of subs

Reply:
Ah, I wasn't clear.

By table, I mean a collection of subscription entries, as we have now, but with no index as we see here.  Instead, we can have well-known keys for each table entry that can be known a priori.

So instead of this:

```
    static StorageKeyName FabricSubscription(FabricIndex fabric, NodeId nodeId)
    {
        return StorageKeyName::Formatted("f/%x/su/%08" PRIX32 "%08" PRIX32, fabric, static_cast<uint32_t>(nodeId >> 32),
                                         static_cast<uint32_t>(nodeId));
    }
    static StorageKeyName SubscriptionResumptionIndex() { return StorageKeyName::FromConst("g/sui"); }
};
```

We can have this:

```
    /**
     * @param subscriptionIndex virtual index in the persisted subscription table
     */
    static StorageKeyName FabricSubscription(uint16_t subscriptionIndex)
    {
        return StorageKeyName::Formatted("f/%x/su/%04" PRIX16, subscriptionIndex);
    }
};
```

And then the persisted index can be nothing more than the maximum subscriptionIndex, making it possible to always deduce and iterate over all possible keys.

Reply:
> It would require quite a large heap buffer to load a big set of subs stored as a single key.

And in fact, this scheme is thriftier in this respect because in the existing approach, you still have a scaling problem with the index itself.

Reply:
I've since revamped the storage API according to Tennessee's iterator suggestion, and the storage format as a flat list according to Michael's suggestion. And so this discussion should be addressed.